tachyon/zk/air/plonky3/base/two_adic_multiplicative_coset.h

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2022 The Plonky3 Authors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.plonky3 and the LICENCE-APACHE.plonky3
+// file.
+
+#ifndef TACHYON_ZK_AIR_PLONKY3_BASE_TWO_ADIC_MULTIPLICATIVE_COSET_H_
+#define TACHYON_ZK_AIR_PLONKY3_BASE_TWO_ADIC_MULTIPLICATIVE_COSET_H_
+
+#include <memory>
+#include <optional>
+#include <utility>
+#include <vector>
+
+#include "tachyon/base/bits.h"
+#include "tachyon/base/containers/container_util.h"
+#include "tachyon/base/optional.h"
+#include "tachyon/math/polynomials/univariate/radix2_evaluation_domain.h"
+#include "tachyon/zk/air/plonky3/base/lagrange_selectors.h"
+
+namespace tachyon::zk::air::plonky3 {
+
+template <typename F>
+class TwoAdicMultiplicativeCoset {
+  static_assert(F::Config::kModulusBits <= 32);
+
+ public:
+  constexpr TwoAdicMultiplicativeCoset() = default;
+
+  TwoAdicMultiplicativeCoset(size_t log_n, const F& shift) {
+    domain_.reset(static_cast<math::Radix2EvaluationDomain<F>*>(
+        math::Radix2EvaluationDomain<F>::Create(1 << log_n)
+            ->GetCoset(shift)
+            .release()));
+  }
+
+  const math::Radix2EvaluationDomain<F>* domain() const {
+    return domain_.get();
+  }
+
+  template <typename ExtField>
+  ExtField GetNextPoint(const ExtField& x) const {
+    return x * domain_->group_gen();
+  }
+
+  TwoAdicMultiplicativeCoset<F> CreateDisjointDomain(size_t min_size) const {
+    return {
+        base::bits::SafeLog2Ceiling(min_size),
+        domain_->offset() * F::FromMontgomery(F::Config::kSubgroupGenerator)};
+  }
+
+  // exp_power_of_2 = x^2^log_n
```

Comment:
I think there should be ().

```suggestion
  // exp_power_of_2 = x^(2^log_n)
```

Reply:
I don't think this comment is actually needed! Thanks for bringing my attention to it~